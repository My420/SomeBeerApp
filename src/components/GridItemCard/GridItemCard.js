import React from 'react';
import PropTypes from 'prop-types';
import styles from './GridItemCard.module.scss';
import GridCardControl from './GridCardControl/GridCardControl';
import GridCardLink from './GridCardLink/GridCardLink';

class GridItemCard extends React.Component {
  state = { isDescriptionOpen: false };

  showDescription = () => {
    this.setState({ isDescriptionOpen: true });
  };

  hideDescription = () => {
    this.setState({ isDescriptionOpen: false });
  };

  toggleDescription = () => {
    const { isDescriptionOpen } = this.state;
    this.setState({ isDescriptionOpen: !isDescriptionOpen });
  };

  addEndListener = (node, done) =>
    node.addEventListener('transitionend', done, false);

  render() {
    const { data } = this.props;
    const { isDescriptionOpen } = this.state;

    return (
      <div className={styles.card}>
        <GridCardLink
          data={data}
          isDescriptionOpen={isDescriptionOpen}
          onCursorEnter={this.showDescription}
          onCursorLeave={this.hideDescription}
        />
        <GridCardControl
          data={data}
          isDescriptionOpen={isDescriptionOpen}
          onToggleButtonClick={this.toggleDescription}
        />
      </div>
    );
  }
}

GridItemCard.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  data: PropTypes.object.isRequired
};

export default GridItemCard;
