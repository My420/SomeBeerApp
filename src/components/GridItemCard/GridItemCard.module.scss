@import '../../sass/variables.scss';

.card {
  font-size: 18px;
  line-height: 24px;
  border: 2px solid $menu-bar-color;
}

.card:hover {
  box-shadow: 0 0 5px 1px $menu-color;
}
