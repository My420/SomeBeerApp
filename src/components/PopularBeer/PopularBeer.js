import React from 'react';
import PropTypes from 'prop-types';
import Immutable from 'immutable';
import { connect } from 'react-redux';
import styles from './PopularBeer.module.scss';
import downloadPopularBeerData from '../../ActionCreator/downloadPopularBeerData';
import ProductList from '../ProductList/ProductList';
import GridItemCard from '../GridItemCard/GridItemCard';
import IconLoadingSvg from '../IconLoadingSvg/IconLoadingSvg';
import ErrorMsg from '../ErrorMsg/ErrorMsg';

export class PopularBeer extends React.Component {
  componentDidMount() {
    const { getBeerData, isLoaded } = this.props;
    if (!isLoaded) {
      getBeerData();
    }
  }

  render() {
    // eslint-disable-next-line no-console
    console.log('render ===== PopularBeer', this.props);

    const { isLoaded, isLoading, isError, errorMessage, data } = this.props;

    return (
      <section className={styles.popular}>
        <h2 className="visually-hidden">Popular Beer</h2>
        {isLoading && (
          <div className={styles.loader}>
            <IconLoadingSvg color="#e69e63" />
          </div>
        )}
        {isError && (
          <div className={styles.error}>
            <ErrorMsg errorMsg={errorMessage} />
          </div>
        )}
        {isLoaded && <ProductList data={data} ProductCard={GridItemCard} />}
      </section>
    );
  }
}

const mapStateToProps = store => {
  return {
    isLoaded: store.popularBeer.isLoaded,
    isLoading: store.popularBeer.isLoading,
    isError: store.popularBeer.isError,
    errorMessage: store.popularBeer.errorMessage,
    data: store.popularBeer.data
  };
};

const mapDispatchToProps = dispatch => {
  return {
    getBeerData: () => {
      dispatch(downloadPopularBeerData());
    }
  };
};

PopularBeer.propTypes = {
  isLoaded: PropTypes.bool.isRequired,
  isLoading: PropTypes.bool.isRequired,
  isError: PropTypes.bool.isRequired,
  errorMessage: PropTypes.string.isRequired,
  data: PropTypes.instanceOf(Immutable.List).isRequired,
  getBeerData: PropTypes.func.isRequired
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(PopularBeer);
