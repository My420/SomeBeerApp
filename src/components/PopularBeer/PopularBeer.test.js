import React from 'react';
import { shallow } from 'enzyme';
import { PopularBeer } from './PopularBeer';
import GridItemCard from '../GridItemCard/GridItemCard';

describe('Test <PopularBeer /> component', () => {
  const props = {
    isLoading: false,
    isLoaded: false,
    isError: false,
    errorMessage: '',
    data: 'test',
    getBeerData: () => {}
  };

  describe('test getBeerData function call', () => {
    test('call the `getBeerData` function in componentDidMount', () => {
      const testFunc = jest.fn();
      const newProps = { ...props, getBeerData: testFunc };
      shallow(<PopularBeer {...newProps} />);

      expect(testFunc.mock.calls.length).toBe(1);
    });

    test('dont call the `getBeerData` function if isLoaded is true', () => {
      const testFunc = jest.fn();
      const newProps = { ...props, isLoaded: true, getBeerData: testFunc };
      shallow(<PopularBeer {...newProps} />);

      expect(testFunc.mock.calls.length).toBe(0);
    });

    test('call the `getBeerData` function if isLoaded is false', () => {
      const testFunc = jest.fn();
      const newProps = { ...props, isError: true, getBeerData: testFunc };
      shallow(<PopularBeer {...newProps} />);

      expect(testFunc.mock.calls.length).toBe(1);
    });
  });

  describe('test <PopularBeer/> component', () => {
    test('should render <IconLoadingSvg/> if isLoading: true', () => {
      const newProps = { ...props, isLoading: true };
      const wrapper = shallow(<PopularBeer {...newProps} />);

      expect(wrapper).toMatchSnapshot();
      expect(wrapper.find('IconLoadingSvg')).toHaveLength(1);
    });

    test('should render <ErrorMsg/> if isError: true', () => {
      const newProps = { ...props, isError: true, errorMessage: 'test error' };
      const wrapper = shallow(<PopularBeer {...newProps} />);

      expect(wrapper).toMatchSnapshot();
      expect(wrapper.find('ErrorMsg')).toHaveLength(1);
    });

    test('should render <PropductList/> if isLoaded: true', () => {
      const newProps = { ...props, isLoaded: true };
      const wrapper = shallow(<PopularBeer {...newProps} />);

      expect(wrapper).toMatchSnapshot();
      expect(wrapper.find('ProductList')).toHaveLength(1);
    });
    test('should correct render <ProductList /> component', () => {
      const newProps = { ...props, isLoaded: true };
      const wrapper = shallow(<PopularBeer {...newProps} />);

      expect(wrapper.find('ProductList').prop('data')).toBe(props.data);
      expect(wrapper.find('ProductList').prop('ProductCard')).toEqual(
        GridItemCard
      );
    });
  });
});
