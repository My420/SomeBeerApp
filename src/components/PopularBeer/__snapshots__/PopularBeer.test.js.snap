// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <PopularBeer /> component test <PopularBeer/> component should render <ErrorMsg/> if isError: true 1`] = `
<section
  className="popular"
>
  <h2
    className="visually-hidden"
  >
    Popular Beer
  </h2>
  <div
    className="error"
  >
    <ErrorMsg
      errorMsg="test error"
    />
  </div>
</section>
`;

exports[`Test <PopularBeer /> component test <PopularBeer/> component should render <IconLoadingSvg/> if isLoading: true 1`] = `
<section
  className="popular"
>
  <h2
    className="visually-hidden"
  >
    Popular Beer
  </h2>
  <div
    className="loader"
  >
    <IconLoadingSvg
      color="#e69e63"
      svgClass="svg"
      wrapperClass="container"
    />
  </div>
</section>
`;

exports[`Test <PopularBeer /> component test <PopularBeer/> component should render <PropductList/> if isLoaded: true 1`] = `
<section
  className="popular"
>
  <h2
    className="visually-hidden"
  >
    Popular Beer
  </h2>
  <ProductList
    ProductCard={[Function]}
    data="test"
  />
</section>
`;
